json.go

```diff
@@ -0,0 +1,797 @@
+            if (obj->fieldNameLengths) {
+                free(obj->fieldNameLengths);
+                obj->fieldNameLengths = NULL;
+            }
+            if (obj->fieldValues) {
+                free(obj->fieldValues);
+                obj->fieldValues = NULL;
+            }
+            break;
+    }
+}
+
+void godror_dpiJsonfreeMem(dpiJsonNode *node) {
+    godror_dpiJsonNodeFree(node);
+    godror_free_dpiNode(node);
+}
+
+*/
+import "C"
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"time"
+	"unsafe"
+)
+
+type JSONOption uint8
+
+var ErrInvalidJSON = errors.New("Invalid JSON Document")
+var ErrInvalidType = errors.New("Invalid JSON Scalar Type")
+
+const (
+	JSONOptDefault        = JSONOption(C.DPI_JSON_OPT_DEFAULT)
+	JSONOptNumberAsString = JSONOption(C.DPI_JSON_OPT_NUMBER_AS_STRING)
+	JSONOptDateAsDouble   = JSONOption(C.DPI_JSON_OPT_DATE_AS_DOUBLE)
+)
+
+// Represents JSON string format, if its in std, ORACLE extended types,
+// BSON extended types, etc
+type JSONStringFlags uint
+
+const (
+	JSONFormatExtnTypes       JSONStringFlags = C.DPI_JSON_USE_EXTENSION_TYPES
+	JSONFormatBSONTypes                       = C.DPI_JSON_BSON_TYPE_PATTERNS
+	JSONFormatBSONTypePattern                 = C.DPI_JSON_USE_BSON_TYPES
+)
+
+// Encapsulates JSON formatted string.
+// it could be reader/writer for streams.
+type JSONString struct {
+	Flags JSONStringFlags // standard , extended types for OSON, BSON
+	Value string          // JSON input
+}
+
+// JSONValue is the interface implemented by structs encapsulating
+// various JSON inputs(map, array, struct(?),scalar values like int, string ..).
+// JSON, JSONObject, JSONArray and JSONScalar implement it.
+type JSONValue interface {
+	// JSONType is Oracle type stored. Its not used now.
+	getOracleJSONType() C.dpiOracleTypeNum
+}
+
+// JSONScalar holds the JSON data to/from Oracle.
+// It includes all scalar values such as int, float, bool, time.Time,
+// time.Duration, byte[] , map and arrays.
+type JSONScalar struct {
+	dpiJsonNode *C.dpiJsonNode
+	isMemOwned  bool // true indicates associated C memory needs to be freed.
+}
+
+// Returns the Go type not the DPI native/oracle type
+func (j JSONScalar) GetValue() (val interface{}, err error) {
+	var d Data
+	jsonNodeToData(&d, j.dpiJsonNode)
+	if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_NUMBER {
+		val, err = getJSONScalarDouble(d)
+	} else if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_VARCHAR {
+		val, err = getJSONScalarString(d)
+	} else {
+		val = d.Get()
+		if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_JSON_OBJECT {
+			jobj := val.(JSONObject)
+			val, err = jobj.GetValue()
+		} else if j.dpiJsonNode.oracleTypeNum == C.DPI_ORACLE_TYPE_JSON_ARRAY {
+			jarr := val.(JSONArray)
+			val, err = jarr.GetValue()
+		} else {
+			err = nil
+		}
+	}
+	return
+}
+
+func (j JSONScalar) getOracleJSONType() C.dpiOracleTypeNum {
+	return j.dpiJsonNode.oracleTypeNum
+}
+
+func (js *JSONScalar) Close() error {
```

Comment:
All members of a struct should use the same reference: either all should be defined on a pointer, or neither.

Current code does not need js to be a pointer,
but if we want Close to be idempotent (don't SIGSEGV on double call),
then we must nil js.dpiJsonNode like
```
dpiJsonNode := js.dpiJsonNode
js.dpiJsonNode = nil
if dpiJsonNode != nil && js.isMemOwner {
  C.godror_dpiJsonfreeMem(dpiJsonNode)
}
return nil
```

The same applies to the other types (JSONArray, JSONValue...)

Reply:
Fixed and updated